File: May-LeetCoding-Challenge/03-Ransom-Note/Ransom-Note.go
// map
func canConstruct(ransomNote string, magazine string) bool {
    letters := make(map[rune]int)
    for _, c := range magazine {
        letters[c]++
    }
    for _, c := range ransomNote {
        letters[c]--
        if letters[c] < 0 {
            return false
        }
    }
    return true
}

// array
func canConstruct(ransomNote string, magazine string) bool {
    var letters [128]int // ASCII table: 128 chars
    for _, c := range magazine {
        letters[c]++
    }
    for _, c := range ransomNote {
        letters[c]--
        if letters[c] < 0 {
            return false
        }
    }
    return true
}

File: May-LeetCoding-Challenge/03-Ransom-Note/Ransom-Note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    var flag = true;
    var stats = new Array(26);
    for(let i = 0;i<26;i++){
        stats[i]=0;
    }
    for(var ch of magazine){
        let index = ch.charCodeAt()-97;//given that 'a'.charCodeAt()=97
        stats[index]++;
    }
    for(var ch of ransomNote){
        let index = ch.charCodeAt()-97;
        stats[index]--;
    }
    for(var int of stats){
        if(int<0){return false;}
    }
    return true;
};
